subsys/bluetooth/mesh/light_hsl_srv.c

```diff
@@ -0,0 +1,549 @@
+	 * server, and instead perform it once at the end of this function.
+	 */
+	srv->pub_pending = true;
+
+	bt_mesh_light_hue_srv_set(&srv->hue, NULL, &set.h, &hue);
+	bt_mesh_light_sat_srv_set(&srv->sat, NULL, &set.s, &sat);
+	lightness_srv_change_lvl(&srv->lightness, ctx, &set.l, &lightness);
+
+	struct bt_mesh_light_hsl_status hsl =
+		HSL_STATUS_INIT(&hue, &sat, &lightness, current);
+
+	if (ack) {
+		hsl_status_rsp(model, ctx, &hsl);
+	}
+
+	srv->pub_pending = false;
+	(void)bt_mesh_light_hsl_srv_pub(srv, NULL, &hsl);
+}
+
+static void hsl_set_handle(struct bt_mesh_model *model,
+			   struct bt_mesh_msg_ctx *ctx,
+			   struct net_buf_simple *buf)
+{
+	hsl_set(model, ctx, buf, true);
+}
+
+static void hsl_set_unack_handle(struct bt_mesh_model *model,
+				 struct bt_mesh_msg_ctx *ctx,
+				 struct net_buf_simple *buf)
+{
+	hsl_set(model, ctx, buf, false);
+}
+
+static void hsl_target_get_handle(struct bt_mesh_model *model,
+				  struct bt_mesh_msg_ctx *ctx,
+				  struct net_buf_simple *buf)
+{
+	struct bt_mesh_light_hsl_srv *srv = model->user_data;
+	struct bt_mesh_lightness_status lightness;
+	struct bt_mesh_light_hue_status hue;
+	struct bt_mesh_light_sat_status sat;
+
+	if (buf->len != BT_MESH_LIGHT_HSL_MSG_LEN_GET) {
+		return;
+	}
+
+	srv->hue.handlers->get(&srv->hue, NULL, &hue);
+	srv->sat.handlers->get(&srv->sat, NULL, &sat);
+	srv->lightness.handlers->light_get(&srv->lightness, NULL, &lightness);
+
+	struct bt_mesh_light_hsl_status status =
+		HSL_STATUS_INIT(&hue, &sat, &lightness, target);
+
+	BT_MESH_MODEL_BUF_DEFINE(rsp, BT_MESH_LIGHT_HSL_OP_TARGET_STATUS,
+				 BT_MESH_LIGHT_HSL_MSG_MAXLEN_STATUS);
+	hsl_status_encode(&rsp, BT_MESH_LIGHT_HSL_OP_TARGET_STATUS, &status);
+
+	bt_mesh_model_send(model, ctx, &rsp, NULL, NULL);
+}
+
+static void default_encode_status(struct net_buf_simple *buf,
+				  struct bt_mesh_light_hsl_srv *srv)
+{
+	bt_mesh_model_msg_init(buf, BT_MESH_LIGHT_HSL_OP_DEFAULT_STATUS);
+	net_buf_simple_add_le16(buf, light_to_repr(srv->lightness.default_light,
+						   ACTUAL));
+	net_buf_simple_add_le16(buf, srv->hue.dflt);
+	net_buf_simple_add_le16(buf, srv->sat.dflt);
+}
+
+static void default_rsp(struct bt_mesh_model *model,
+			struct bt_mesh_msg_ctx *rx_ctx)
+{
+	struct bt_mesh_light_hsl_srv *srv = model->user_data;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_LIGHT_HSL_OP_DEFAULT_STATUS,
+				 BT_MESH_LIGHT_HSL_MSG_LEN_DEFAULT);
+	default_encode_status(&msg, srv);
+	(void)bt_mesh_model_send(model, rx_ctx, &msg, NULL, NULL);
+}
+
+static void default_set(struct bt_mesh_model *model,
+			struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
+{
+	struct bt_mesh_light_hsl_srv *srv = model->user_data;
+
+	lightness_srv_default_set(&srv->lightness, ctx,
+				  repr_to_light(net_buf_simple_pull_le16(buf),
+						ACTUAL));
+	bt_mesh_light_hue_srv_default_set(&srv->hue, ctx,
+					      net_buf_simple_pull_le16(buf));
+	bt_mesh_light_sat_srv_default_set(&srv->sat, ctx,
+					      net_buf_simple_pull_le16(buf));
```

Comment:
Would it be good to create a decoder for this into a "bt_mesh_light_hsl" and then use that for the set functions.